tikv/raftstore/fsm.go

```diff
@@ -0,0 +1,556 @@
+
+	/// Set a mailbox to fsm, which should be used to send message to itself.
+	setMailbox(mb *mailbox)
+	/// Take the mailbox from fsm. Implementation should ensure there will be
+	/// no reference to mailbox after calling this method.
+	takeMailbox() *mailbox
+}
+
+type fsmScheduler interface {
+	/// Schedule a fsm for later handles.
+	schedule(fsm fsm)
+	/// Shutdown the scheduler, which indicates that resouces like
+	/// background thread pool should be released.
+	shutdown()
+}
+
+var _ fsmScheduler = new(normalScheduler)
+
+type normalScheduler struct {
+	sender chan<- *Msg
+}
+
+func (s *normalScheduler) schedule(fsm fsm) {
+	s.sender <- &Msg{
+		Type: MsgTypeFsmNormal,
+		Fsm:  fsm,
+	}
+}
+
+func (s *normalScheduler) shutdown() {
+	close(s.sender)
+}
+
+var _ fsmScheduler = new(controlScheduler)
+
+type controlScheduler struct {
+	sender chan<- *Msg
+}
+
+func (s *controlScheduler) schedule(fsm fsm) {
+	s.sender <- &Msg{
+		Type: MsgTypeFsmControl,
+		Fsm:  fsm,
+	}
+}
+
+func (s *controlScheduler) shutdown() {
+	// We don't close the sender here because it will be closed by normalScheduler.
+}
+
+const (
+	notifyStateNotified = 0
+	notifyStateIdle     = 1
+	notifyStateDrop     = 2
+)
+
+type fsmState struct {
+	status int64
+	data   unsafe.Pointer
+}
+
+/// Every mailbox should have one and only one owner, who will receive all
+/// messages sent to this mailbox.
+///
+/// When a message is sent to a mailbox, its owner will be checked whether it's
+/// idle. An idle owner will be scheduled via `fsmScheduler` immediately, which
+/// will drive the fsm to poll for messages.
+type mailbox struct {
+	sender chan<- *Msg
+	state  fsmState
+}
+
+func newMailbox(sender chan<- *Msg, fsm fsm) *mailbox {
+	mb := &mailbox{sender: sender}
+	mb.state.status = notifyStateIdle
+	mb.state.data = unsafe.Pointer(&fsm)
+	return mb
+}
+
+/// Take the owner if it's IDLE.
+func (mb *mailbox) takeFsm() fsm {
+	swapped := atomic.CompareAndSwapInt64(&mb.state.status, notifyStateIdle, notifyStateNotified)
+	if !swapped {
+		return nil
+	}
+	old := atomic.SwapPointer(&mb.state.data, nil)
+	if old == nil {
+		panic("inconsistent status and data, something should be wrong.")
+	}
+	return *(*fsm)(old)
+}
+
+/// Put the owner back to the state.
+///
+/// It's not required that all messages should be consumed before
+/// releasing a fsm. However, a fsm is guaranteed to be notified only
+/// when new messages arrives after it's released.
+func (mb *mailbox) release(fsm fsm) {
+	prev := atomic.SwapPointer(&mb.state.data, (unsafe.Pointer)(&fsm))
+	prevStatus := int64(notifyStateNotified)
+	if prev == nil {
+		swapped := atomic.CompareAndSwapInt64(&mb.state.status, notifyStateNotified, notifyStateIdle)
+		if swapped {
+			return
+		}
+		prevStatus = atomic.LoadInt64(&mb.state.status)
+		if prevStatus == notifyStateDrop {
+			atomic.SwapPointer(&mb.state.data, nil)
+		}
+	}
+	panic(fmt.Sprintf("invalid release state %d", prevStatus))
+}
+
+func (mb *mailbox) len() int {
```

Comment:
Is there `inline` key words?

Reply:
Go compiler will inline simply code automatically.